File: leetcode.com/0990_satisfiability_of_equality_equations/solution.c
#define CHARS_COUNT 26

int p[CHARS_COUNT];

int find(int x)
{
    if (p[x] == x)
        return x;
    return p[x] = find(p[x]);
}

void _union(int x, int y)
{
    p[find(x)] = find(y);
}

bool equationsPossible(char **equations, int equationsSize)
{
    for (int i = 0 ; i < CHARS_COUNT ; i++)
        p[i] = i;

    /* assign */
    for (int i = 0 ; i < equationsSize ; i++) {
        if (equations[i][1] == '!')
            continue;

        int x = equations[i][0] - 'a';
        int y = equations[i][3] - 'a';
        _union(y, x);
    }

    /* check */
    for (int i = 0 ; i < equationsSize ; i++) {
        if (equations[i][1] == '=')
            continue;

        int x = equations[i][0] - 'a';
        int y = equations[i][3] - 'a';

        if (find(x) == find(y))
            return false;
    }

    return true;
}

File: leetcode.com/0990_satisfiability_of_equality_equations/solution.py
class UnionFind:

    def __init__(self, n):
        self.p = list(range(n))

    def union(self, x, y):
        x, y = self.find(x), self.find(y)
        self.p[x] = y

    def find(self, x):
        if self.p[x] == x:
            return x

        self.p[x] = self.find(self.p[x])
        return self.p[x]

    def connected(self, x, y):
        return self.find(x) == self.find(y)


class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        uf = UnionFind(26)
        equations.sort(key=lambda t: t[1] == '!')

        for eq in equations:
            x, y, e = ord(eq[0]) - ord('a'), ord(eq[3]) - ord('a'), eq[1] == '='

            if e:
                uf.union(x, y)
            elif uf.connected(x, y):
                return False

        return True
